File: week-03/day-2/src/piratelife/Ship.java
package piratelife;

import java.util.ArrayList;
import java.util.List;

public class Ship {
    List<Pirate> crew = new ArrayList<>();
    Pirate captain;
    String name;

    Ship(String name) {
        this.name = name;
    }

    public int getRandom(int maxRandom) {
        maxRandom = (int) ((Math.random()) * maxRandom);
        return maxRandom;
    }

    public void fillShip(int howManyPiratesCanFit) {
        captain = new Pirate("Barbosa", 40);
        int numberOfPirates = getRandom(howManyPiratesCanFit);
        for (int i = 0; i < numberOfPirates; i++) {
            Pirate jack = new Pirate("Jack" + i, 20 + i);
            this.crew.add(jack);
        }
    }

    public int countLivingPirates() {
        int numberOfLivingPirates = 0;
        for (Pirate pirate : crew) {
            if (pirate.alive) {
                numberOfLivingPirates++;
            }
        }
        return numberOfLivingPirates;
    }


    //the ship should win if its calculated score is higher
    //calculate score: Number of Alive pirates in the crew - Number of consumed rum by the captain
    public boolean battle(Ship otherShip) {
        boolean won;
        // if this ship won, set won to true, kill pirates from other ship, this ship partyHard
        if (countLivingPirates() - captain.intoxication > otherShip.countLivingPirates() - otherShip.captain.intoxication) {
            won = true;
            removeRandomNumberOfPirates(otherShip);
            partyHard(this);

            // if other ship won, set won to false, kill pirates from this ship, Other ship partyHard
        } else {
            won = false;
            removeRandomNumberOfPirates(this);
            partyHard(otherShip);
        }
        return won;
    }

    //kill random number of ship crew, **used for ship.battle**
    private void removeRandomNumberOfPirates(Ship ship) {

        if (getRandom(6) < 2) {
            captain.die();
        }
        for (int i = 0; i < (countLivingPirates() - getRandom(countLivingPirates())); i++) {
            ship.crew.remove(i);
        }
    }

    // each pirate from ship will drink random number of rums (max 20)
    private void partyHard(Ship ship) {
        for (int i = 0; i < getRandom(20); i++) {
            ship.captain.drinkSomeRum();
        }
        for (Pirate pirate : ship.crew) {
            for (int i = 0; i < getRandom(20); i++) {
                pirate.drinkSomeRum();
            }
        }
    }

    @Override
    public String toString() {
        String shipCrew = "Captain of " + name + " is " + captain.name + ". He consumed " + captain.intoxication + " rums. Alive? " + captain.alive + "\n";
        shipCrew += "There are " + countLivingPirates() + " living crew members on this ship";
        return shipCrew;
    }
}

File: Python/PirateApp/Ship.py
from Pirate import Pirate
import random


class Ship:

    def __init__(self):
        self.pirates = []
        self.captain = ""
        self.fill_ship()

    def fill_ship(self):
        self.number_of_pirates = round(random.random()*50)
        self.captain = Pirate(True)
        for x in range (self.number_of_pirates):
            self.pirates.append(Pirate())

    def get_pirates_alive(self):
        count = 0
        for pirate in self.pirates:
            if(pirate.is_dead == False):
                count +=1
        return count

    def __str__(self):
        status ="Captain intoxication: {}, Dead: {}, alive crew: {}".format(self.captain.intoxication,self.captain.is_dead, self.get_pirates_alive())
        return status

    def battle (self, another_ship):
        this_ship_power = self.get_battle_power(self)
        another_ship_power = self.get_battle_power(another_ship)
        if this_ship_power>another_ship_power:
            self.__win(self)
            self.__loose(another_ship)
            return True
        else:
            self.__win(another_ship)
            self.__loose(self)
            return False

    def get_battle_power (self, ship):
        return ship.get_pirates_alive() - ship.captain.intoxication

    def __loose (self, ship):
        options = "live","die"
        destiny = random.choice(options)
        if destiny=="die":
            ship.captain.die()
        for pirate in ship.pirates:
            destiny = random.choice(options)
            if destiny=="die":
                pirate.die()
            
    def __win (self, ship):
        rum_ammount = round(random.random()*50)
        ship.captain.intoxication += rum_ammount
        for pirate in ship.pirates:
            pirate.intoxication += rum_ammount
